airbyte-integrations/connectors/source-intercom/source_intercom/source.py

```diff
@@ -0,0 +1,364 @@
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+from abc import ABC
+from datetime import datetime
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple, Union
+from urllib.parse import parse_qsl, urlparse
+
+import requests
+from airbyte_cdk.logger import AirbyteLogger
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth import HttpAuthenticator, TokenAuthenticator
+
+
+class IntercomStream(HttpStream, ABC):
+    url_base = "https://api.intercom.io/"
+
+    # https://developers.intercom.com/intercom-api-reference/reference#rate-limiting
+    queries_per_hour = 1000  # 1000 queries per hour == 1 req in 3,6 secs
+
+    primary_key = "id"
+    data_fields = ["data"]
+
+    def __init__(
+        self,
+        authenticator: HttpAuthenticator,
+        start_date: str = None,
+        **kwargs,
+    ):
+        self.start_date = start_date
+
+        super().__init__(authenticator=authenticator)
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        """
+        Abstract method of HttpStream - should be overwritten.
+        Returning None means there are no more pages to read in response.
+        """
+
+        next_page = response.links.get("next", None)
+
+        if next_page:
+            return dict(parse_qsl(urlparse(next_page.get("url")).query))
+        else:
+            return None
+
+    def request_headers(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> Mapping[str, Any]:
+        return {"Accept": "application/json"}
+
+    def _send_request(self, request: requests.PreparedRequest, request_kwargs: Mapping[str, Any]) -> requests.Response:
+        try:
+            return super()._send_request(request, request_kwargs)
+        except requests.exceptions.HTTPError as e:
+            error_message = e.response.text
+            if error_message:
+                self.logger.error(f"Stream {self.name}: {e.response.status_code} " f"{e.response.reason} - {error_message}")
+            raise e
+
+    def get_data(self, response: requests.Response) -> List:
```

Comment:
just a small thing but we can probably merge this into `parse_response` and then in the incremental stream override of `parse_response` we can do `super().parse_response()` then mutate the output of that method

Reply:
Done